File: src/year2025/day6/day6.py
#!/usr/bin/env python3
# [Day 6: Trash Compactor](https://adventofcode.com/2025/day/6)


import atexit
import time
from argparse import ArgumentParser
from functools import reduce
from operator import mul
from pathlib import Path

parser = ArgumentParser()
parser.add_argument("-v", "--verbose", action="store_true")
parser.add_argument("-t", "--test", action="store_true")
parser.add_argument("--elapsed", action="store_true")
parser.add_argument("filename", nargs="?", type=Path, default="input.txt")
args = parser.parse_args()
if args.test:
    args.filename = Path("test.txt")

data = args.filename.read_text()

if args.elapsed:
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


# Part 1

table = []
for line in data.splitlines():
    if "+" in line:
        operators = list(line.split())
    else:
        table.append(list(map(int, line.split())))

part1 = 0
for i, op in enumerate(operators):
    if op == "+":
        s = 0
        for row in table:
            s += row[i]
    elif op == "*":
        s = 1
        for row in table:
            s *= row[i]
    part1 += s
print(part1)


# Part 2

table = []
for line in data.splitlines():
    if "+" in line:
        operators = line
    else:
        table.append(line)

part2 = 0
for p, op in enumerate(operators):
    if op in "+*":
        q = p + 1
        while q < len(operators) and operators[q] == " ":
            q += 1
        width = q - p
        if q < len(operators):
            width -= 1

        nums = []
        for i in range(p, p + width):
            vn = 0
            for row in table:
                if row[i] != " ":
                    vn = vn * 10 + int(row[i])
            nums.append(vn)

        if op == "+":
            part2 += sum(nums)
        elif op == "*":
            part2 += reduce(mul, nums)
print(part2)

File: src/year2025/day6/day6.rs
//! [Day 6: Trash Compactor](https://adventofcode.com/2025/day/6)

struct Puzzle<'a> {
    //
    rows: Vec<&'a str>,
    operators: &'a str,
}

impl<'a> Puzzle<'a> {
    /// Initialize from the puzzle input.
    fn new(data: &'a str) -> Self {
        let mut rows = vec![];
        let mut operators = "";

        for line in data.lines() {
            if line.contains('+') || line.contains('*') {
                operators = line;
            } else {
                rows.push(line);
            }
        }

        Self { rows, operators }
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        let mut result = 0;
        let rows = self
            .rows
            .iter()
            .map(|row| row.split_ascii_whitespace().collect::<Vec<&str>>())
            .collect::<Vec<Vec<&str>>>();

        for (i, op) in self.operators.split_ascii_whitespace().enumerate() {
            result += match op {
                "+" => rows.iter().map(|row| row[i].parse::<u64>().unwrap()).sum(),
                "*" => rows
                    .iter()
                    .map(|row| row[i].parse::<u64>().unwrap())
                    .product::<u64>(),
                _ => panic!("unknown op {op}"),
            };
        }
        result
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        let mut result = 0;

        for (p, op) in self.operators.char_indices().filter(|(_, op)| op != &' ') {
            let mut end = p + 1;
            while end < self.operators.len() && self.operators.chars().nth(end) == Some(' ') {
                end += 1;
            }
            if end != self.operators.len() {
                end -= 1;
            }

            let mut nums = vec![];

            for i in p..end {
                let mut vnum = 0;
                for row in &self.rows {
                    if let Some(ch) = row.chars().nth(i)
                        && ch != ' '
                    {
                        vnum = vnum * 10 + u64::from(ch.to_digit(10).unwrap());
                    }
                }
                nums.push(vnum);
            }

            result += match op {
                '+' => nums.iter().sum(),
                '*' => nums.iter().product::<u64>(),
                _ => panic!("unknown op {op}"),
            };
        }

        result
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn part1() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 4277556);
    }

    #[test]
    fn part2() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 3263827);
    }
}
